### PR TITLE
[SYCL] Add SYCL_HOST_UNIFIED_MEMORY environment variable

### PR DESCRIPTION
**Pull request by sergey-semenov**
Add the SYCL_HOST_UNIFIED_MEMORY environment variable that forces the
execution graph builder to behave as if either no non-host devices support
host unified memory or all of them do. The purpose of this variable is to
facilitate testing and debugging of host unified memory specific
behaviour in the runtime library.

### CODE REVIEW

File: sycl/doc/EnvironmentVariables.md [hunk 1]
```diff
@@ -33,6 +33,7 @@ subject to change. Do not rely on these variables in production code.
 | SYCL_PARALLEL_FOR_RANGE_ROUNDING_TRACE | Any(\*) | Enables tracing of parallel_for invocations with rounded-up ranges. |
 | SYCL_DISABLE_PARALLEL_FOR_RANGE_ROUNDING | Any(\*) | Disables automatic rounding-up of parallel_for invocation ranges. |
 | SYCL_ENABLE_PCI | Integer | When set to 1, enables obtaining the GPU PCI address when using the Level Zero backend. The default is 0. |
+| SYCL_HOST_UNIFIED_MEMORY | Integer | Force the execution graph builder to behave as if no non-host devices support host unified memory when set to 0, or as if all of them do when set to 1. |
```
Comment (pvchupin) [inline sycl/doc/EnvironmentVariables.md:36]:
```suggestion
| SYCL_HOST_UNIFIED_MEMORY | Integer | Enforce host unified memory support/unsupport for the execution graph builder. If set to 0 it is enforced as not supported by all devices. If set to 1 it is enforced as supported by all devices. |
```

Comment (kbobrovs) [reply sycl/doc/EnvironmentVariables.md:36]: As a variant:
`Control host unified memory support for the execution graph builder. Setting to 0/1 forces the graph builder to unsupport/support the host USM on all devices.`

**Review by pvchupin — APPROVED** [review]
doc ok

**Comment by sergey-semenov** [comment]
@rbegam @intel/llvm-reviewers-runtime Could someone please take a look?

File: sycl/source/detail/scheduler/graph_builder.cpp [hunk 2]
```diff
@@ -597,6 +598,12 @@ Scheduler::GraphBuilder::findAllocaForReq(MemObjRecord *Record,
 }
 
 static bool checkHostUnifiedMemory(const ContextImplPtr &Ctx) {
+  if (const char *HUMConfig = SYCLConfig<SYCL_HOST_UNIFIED_MEMORY>::get()) {
+    if (std::strcmp(HUMConfig, "0") == 0)
+      return Ctx->is_host();
+    if (std::strcmp(HUMConfig, "1") == 0)
+      return true;
```
Comment (alexbatashev) [inline sycl/source/detail/scheduler/graph_builder.cpp:605]: I'd prefer `strncmp` here

**Review by alexbatashev — APPROVED** [review]
Just a small nit, otherwise LGTM

File: sycl/source/detail/scheduler/graph_builder.cpp (hunk 2 above)
Comment (sergey-semenov) [reply sycl/source/detail/scheduler/graph_builder.cpp:605]: I would prefer to keep it as `strcmp` to have only 0/1 as valid values and prevent something like `SYCL_HOST_UNIFIED_MEMORY=01 ./a.out` from having an effect.

Comment (alexbatashev) [reply sycl/source/detail/scheduler/graph_builder.cpp:605]: In the config you explicitly specified 1 byte to be used for storage. So, in that case only the first character should be read.

Comment (sergey-semenov) [reply sycl/source/detail/scheduler/graph_builder.cpp:605]: The 1 byte storage is only used to store a configuration value if it is set via a config file, it's completely bypassed if an environment variable is used instead. So in the example I gave above, the config value is not truncated.

**Pull request merged**